src/server/rdb_save.cc

```diff
@@ -1044,75 +1090,43 @@ void RdbSaver::Cancel() {
   }
-  ~ZstdCompressImpl() {
-    ZSTD_freeCCtx(cctx_);
 
-    VLOG(1) << "zstd compressed size: " << compressed_size_total_;
-    VLOG(1) << "zstd uncompressed size: " << uncompressed_size_total_;
+  // Compress the data
+  if (!compressor_impl_) {
+    compressor_impl_.reset(new ZstdCompressImpl());
   }
 
-  std::string_view Compress(std::string_view str);
-
- private:
-  ZSTD_CCtx* cctx_;
-  int compression_level_ = 1;
-  base::PODArray<uint8_t> compr_buf_;
-  uint32_t compressed_size_total_ = 0;
-  uint32_t uncompressed_size_total_ = 0;
-};
-
-std::string_view ZstdCompressSerializer::ZstdCompressImpl::Compress(string_view str) {
-  size_t buf_size = ZSTD_compressBound(str.size());
-  if (compr_buf_.size() < buf_size) {
-    compr_buf_.reserve(buf_size);
-  }
-  size_t compressed_size = ZSTD_compressCCtx(cctx_, compr_buf_.data(), compr_buf_.capacity(),
-                                             str.data(), str.size(), compression_level_);
-
-  compressed_size_total_ += compressed_size;
-  uncompressed_size_total_ += str.size();
-  return string_view(reinterpret_cast<const char*>(compr_buf_.data()), compressed_size);
-}
-
-ZstdCompressSerializer::ZstdCompressSerializer() {
-  impl_.reset(new ZstdCompressImpl());
-}
-
-optional<string> ZstdCompressSerializer::Compress(std::string_view str) {
-  if (str.size() < kMinStrSizeToCompress) {
-    ++small_str_count_;
-    return nullopt;
+  Bytes compressed_blob = compressor_impl_->Compress(blob_to_compress);
+  if (compressed_blob.length() > blob_size * kMinCompressionReductionPrecentage) {
+    ++compression_stats_.compression_no_effective;
+    return;
   }
 
-  // Compress the string
-  string_view compressed_res = impl_->Compress(str);
-  if (compressed_res.size() > str.size() * kMinCompressionReductionPrecentage) {
-    ++compression_no_effective_;
-    return nullopt;
-  }
+  // Clear membuf and write the compressed blob to it
+  mem_buf_.ConsumeInput(blob_size);
 
-  string serialized_compressed_blob;
   // First write opcode for compressed string
-  serialized_compressed_blob.push_back(RDB_OPCODE_COMPRESSED_BLOB_START);
-  // Get compressed string len encoded
-  uint8_t buf[9];
-  unsigned enclen = SerializeLen(compressed_res.size(), buf);
-
-  // Write encoded compressed string len and than the compressed string
-  serialized_compressed_blob.append(reinterpret_cast<const char*>(buf), enclen);
-  serialized_compressed_blob.append(compressed_res);
-  return serialized_compressed_blob;
-}
-
-ZstdCompressSerializer::~ZstdCompressSerializer() {
-  VLOG(1) << "zstd compression not effective: " << compression_no_effective_;
-  VLOG(1) << "small string none compression applied: " << small_str_count_;
+  auto dest = mem_buf_.AppendBuffer();
+  dest[0] = RDB_OPCODE_COMPRESSED_BLOB_START;
+  mem_buf_.CommitWrite(1);
+
+  // Write encoded compressed blob len
+  dest = mem_buf_.AppendBuffer();
+  unsigned enclen = SerializeLen(compressed_blob.length(), dest.data());
+  mem_buf_.CommitWrite(enclen);
+
```

Comment:
So kMinCompressionReductionPrecentage and kMinStrSizeToCompress should guarantee, that if we write opcode + size + compressed blob, we still fit into the previous buffer without reserving capacity

Reply:
But this requires pretty bad configs, for example percentage 99% and min size 128 can lead to only two bytes being saved, but three written (opcode + 2 byte len). Not sure if this needs a CHECK or not, because it's hard to cause, but also hard to find in case of a bug

Reply:
Ohh no I do not want to make such assumptions , I will call reserve